### PR TITLE
perf: cache dynamic imports of nodes

### PR DESCRIPTION
**Pull request by gtm-nayan**
Dunno why but node spends a lot of time in esm loader land even for modules that have already been imported. This PR makes it so that the dynamic imports are only called once and that value is cached for any future calls. This increases rps by a little over 1.5x on my machine.

Before:
image

After:
image

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [ ] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Comment by Rich-Harris** [comment]
Is there any reason we wouldn't _always_ cache loaders? There's no danger of a memory leak since you can't clear the module cache anyway

**Comment by gtm-nayan** [comment]
I was thinking of dev mode and hosts which would create a new environment for each request.

**Comment by benmccann** [comment]
I was curious based off your comments on Discord, would we implement this differently if we could make breaking changes? just wondering if we should add a TODO for kit v2?

**Comment by gtm-nayan** [comment]
Maybe? Not necessarily Kit v2 but it'd be a breaking change
I was think of adding something like `injectImport` to builder.generateManifest params and then using that to pull the __memo function from `@sveltejs/kit/private-utils`, same for all the component wrappers.

But thinking about it more, that isn't much cleaner so I'd likely leave it as is. Unless a better solution comes along.

File: packages/kit/src/exports/index.js [hunk 1]
```diff
@@ -84,7 +84,10 @@ const encoder = new TextEncoder();
 export function text(body, init) {
 	const headers = new Headers(init?.headers);
 	if (!headers.has('content-length')) {
-		headers.set('content-length', encoder.encode(body).byteLength.toString());
+		const encoded = encoder.encode(body);
+		headers.set('content-length', encoded.byteLength.toString());
+		// @ts-expect-error
```
Comment (benmccann) [inline packages/kit/src/exports/index.js:89]: I think we should be able to write this without the `@ts-expect-error`?

File: packages/kit/src/exports/index.js [hunk 2]
```diff
@@ -84,7 +84,10 @@ const encoder = new TextEncoder();
...
+		body = encoded; // reuse encoded value to avoid duplicating work inside Response
```
Comment (gtm-nayan) [inline packages/kit/src/exports/index.js:90]:
```suggestion
		return new Response(encoded, {
				...init,
				headers
		})
```

File: packages/kit/src/core/generate_manifest/index.js [hunk 3]
```diff
@@ -65,8 +65,8 @@ export function generate_manifest({ build_data, relative_path, routes }) {
 
 	// prettier-ignore
 	// String representation of
-	/** @type {import('@sveltejs/kit').SSRManifest} */
-	return dedent`
+	/** @template {import('@sveltejs/kit').SSRManifest} T */
```
Comment (benmccann) [inline packages/kit/src/core/generate_manifest/index.js:68]: I don't think it's correct to use `@template` here. That's for type parameters and not template literals: https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#template. Unless maybe I'm missing where the `T` is?

Comment (gtm-nayan) [reply packages/kit/src/core/generate_manifest/index.js:68]: it's a hack to get VSCode to create a jump link for the SSRManifest. setting it as @type gave an error because the string is not the type SSRManifest

File: packages/kit/src/exports/index.js [hunk 4]
```diff
@@ -84,7 +84,12 @@ const encoder = new TextEncoder();
 export function text(body, init) {
 	const headers = new Headers(init?.headers);
 	if (!headers.has('content-length')) {
-		headers.set('content-length', encoder.encode(body).byteLength.toString());
+		const encoded = encoder.encode(body);
+		headers.set('content-length', encoded.byteLength.toString());
+		return new Response(encoded, {
```
Comment (dummdidumm) [inline packages/kit/src/exports/index.js:89]: is this a drive-by fix?

File: packages/kit/src/core/generate_manifest/index.js [hunk 5]
```diff
@@ -44,7 +44,7 @@ export function generate_manifest({ build_data, relative_path, routes }) {
 	);
 
 	/** @type {(path: string) => string} */
-	const loader = (path) => `() => import('${path}')`;
+	const loader = (path) => `__memo(() => import('${path}'))`;
```
Comment (dummdidumm) [inline packages/kit/src/core/generate_manifest/index.js:47]: Small comment maybe why we memo the import

File: packages/kit/src/core/generate_manifest/index.js [hunk 6]
```diff
@@ -97,10 +97,24 @@ export function generate_manifest({ build_data, relative_path, routes }) {
...
+						matchers, 
+						type => `const { match: ${type} } = await import ('${(join_relative(relative_path, `/entries/matchers/${type}.js`))}')`
+					).join('\n')}
 					return { ${Array.from(matchers).join(', ')} };
 				}
 			}
 		}
 	`;
+
+	return dedent`
```
Comment (dummdidumm) [inline packages/kit/src/core/generate_manifest/index.js:110]: Small comment maybe why we need the memo

**Review by dummdidumm — COMMENTED** [review]
Anything else that needs to be adressed here (besides my comment suggestions) or is this good to go?

File: packages/kit/src/exports/index.js (hunk 4 above)
Comment (gtm-nayan) [reply packages/kit/src/exports/index.js:89]: Another perf thing to avoid duplicating the encoding work inside the Response constructor.

File: packages/kit/src/core/generate_manifest/index.js [hunk 7]
```diff
@@ -97,10 +97,24 @@ export function generate_manifest({ build_data, relative_path, routes }) {
...
+					${Array.from(
+						matchers, 
+						type => `const { match: ${type} } = await import ('${(join_relative(relative_path, `/entries/matchers/${type}.js`))}')`
+					).join('\n')}
 					return { ${Array.from(matchers).join(', ')} };
 				}
 			}
 		}
 	`;
+
```
Comment (gtm-nayan) [inline packages/kit/src/core/generate_manifest/index.js:109]:
```suggestion

	// Memoize the loaders to prevent Node from doing unnecessary work
	// on every dynamic import call	
```

File: packages/kit/src/core/generate_manifest/index.js (hunk 5 above)
Comment (gtm-nayan) [reply packages/kit/src/core/generate_manifest/index.js:47]: Added a comment near the __memo declaration below. That should suffice.

**Review by dummdidumm — APPROVED** [review]
